File: v1/code/go/src/jvmgo/ch06/rtda/operand_stack.go
package rtda

import "math"
import "jvmgo/ch06/rtda/heap"

type OperandStack struct {
	size  uint
	slots []Slot
}

func newOperandStack(maxStack uint) *OperandStack {
	if maxStack > 0 {
		return &OperandStack{
			slots: make([]Slot, maxStack),
		}
	}
	return nil
}

func (self *OperandStack) PushInt(val int32) {
	self.slots[self.size].num = val
	self.size++
}
func (self *OperandStack) PopInt() int32 {
	self.size--
	return self.slots[self.size].num
}

func (self *OperandStack) PushFloat(val float32) {
	bits := math.Float32bits(val)
	self.slots[self.size].num = int32(bits)
	self.size++
}
func (self *OperandStack) PopFloat() float32 {
	self.size--
	bits := uint32(self.slots[self.size].num)
	return math.Float32frombits(bits)
}

// long consumes two slots
func (self *OperandStack) PushLong(val int64) {
	self.slots[self.size].num = int32(val)
	self.slots[self.size+1].num = int32(val >> 32)
	self.size += 2
}
func (self *OperandStack) PopLong() int64 {
	self.size -= 2
	low := uint32(self.slots[self.size].num)
	high := uint32(self.slots[self.size+1].num)
	return int64(high)<<32 | int64(low)
}

// double consumes two slots
func (self *OperandStack) PushDouble(val float64) {
	bits := math.Float64bits(val)
	self.PushLong(int64(bits))
}
func (self *OperandStack) PopDouble() float64 {
	bits := uint64(self.PopLong())
	return math.Float64frombits(bits)
}

func (self *OperandStack) PushRef(ref *heap.Object) {
	self.slots[self.size].ref = ref
	self.size++
}
func (self *OperandStack) PopRef() *heap.Object {
	self.size--
	ref := self.slots[self.size].ref
	self.slots[self.size].ref = nil
	return ref
}

func (self *OperandStack) PushSlot(slot Slot) {
	self.slots[self.size] = slot
	self.size++
}
func (self *OperandStack) PopSlot() Slot {
	self.size--
	return self.slots[self.size]
}

File: v1/code/java/jvmgo_java/ch05/src/main/java/com/github/jvmgo/rtda/OperandStack.java
package com.github.jvmgo.rtda;

import lombok.Getter;
import lombok.Setter;
import lombok.ToString;

import java.lang.ref.Reference;
import java.util.Stack;

@Getter
@Setter
@ToString
public class OperandStack {

    private Stack<Slot> stacks;

    private int maxStack;

    public OperandStack(int maxStack) {
        this.maxStack = maxStack;
        if (maxStack > 0) {
            stacks = new Stack<>();
        } else
            System.exit(-1);
    }

    public void pushSlot(Slot slot) throws Exception {
        if (stacks.size() >= maxStack) {
            throw new Exception("size > maxStack");
        }
        stacks.push(slot);
    }

    public Slot popSlot() throws Exception {
        if (stacks.size() <= 0) {
            throw new Exception("size <0 ");
        }
        return stacks.pop();
    }

    public void pushInt(int val) throws Exception {
        pushSlot(Util.setInt(val));
    }

    public int popInt() {
        return Util.getInt(stacks.pop());
    }

    public void pushFloat(float val) throws Exception {
        int a = Float.floatToIntBits(val);
        pushInt(a);
    }

    public float popFloat() {
        return Float.intBitsToFloat(popInt());
    }

    public void pushLong(long val) throws Exception {
        Slot[] slots = Util.setLong(val);
        pushSlot(slots[0]);
        pushSlot(slots[1]);
    }

    public long popLong() throws Exception {
        Slot low = popSlot();
        Slot high = popSlot();
        return Util.getLong(new Slot[]{high, low});
    }

    public void pushDouble(double val) throws Exception {
        pushLong(Double.doubleToLongBits(val));
    }

    public double popDouble() throws Exception {
        return Double.longBitsToDouble(popLong());
    }

    public void pushRef(Reference ref) throws Exception {
        Slot slot = new Slot();
        slot.setRef(ref);
        pushSlot(slot);
    }

    public Reference popRef() throws Exception {
        Reference reference = popSlot().getRef();
        return reference;
    }
}
